### ISSUE TITLE
Git generator - Error fetching repo with GitHub app

### ISSUE THREAD

**Issue by agaudreault-jive** [argoproj/applicationset#480]
I have a GitHub app credentials configured globally in ArgoCD for all repository in `my-org`. However, when an application use the git generator, if fails every now and then to fetch the repository with the error `could not get access_tokens from GitHub API for installation ID <REDACTED>: context deadline exceeded`

Error:
```
Reconciler error
could not refresh installation id <REDACTED>'s token: could not get access_tokens from GitHub API for installation ID <REDACTED>: context deadline exceeded
Error during fetching repo
github.com/argoproj-labs/applicationset/pkg/services.checkoutRepo
	/workspace/pkg/services/repo_service.go:137
github.com/argoproj-labs/applicationset/pkg/services.(*argoCDService).GetDirectories
	/workspace/pkg/services/repo_service.go:87
github.com/argoproj-labs/applicationset/pkg/generators.(*GitGenerator).generateParamsForGitDirectories
	/workspace/pkg/generators/git.go:74
github.com/argoproj-labs/applicationset/pkg/generators.(*GitGenerator).GenerateParams
	/workspace/pkg/generators/git.go:58
github.com/argoproj-labs/applicationset/pkg/generators.Transform
	/workspace/pkg/generators/generator_spec_processor.go:52
github.com/argoproj-labs/applicationset/pkg/controllers.(*ApplicationSetReconciler).generateApplications
	/workspace/pkg/controllers/applicationset_controller.go:432
github.com/argoproj-labs/applicationset/pkg/controllers.(*ApplicationSetReconciler).Reconcile
	/workspace/pkg/controllers/applicationset_controller.go:96
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.0/pkg/internal/controller/controller.go:298
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.0/pkg/internal/controller/controller.go:253
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.0/pkg/internal/controller/controller.go:214
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1371
```

Secret configuration:
```yaml
apiVersion: v1
kind: Secret
type: Opaque
metadata:
  name: github-app
  labels:
    argocd.argoproj.io/instance: argo-cd
    argocd.argoproj.io/secret-type: repo-creds
data:
  githubAppID: <REDACTED>
  githubAppInstallationID: <REDACTED>
  githubAppPrivateKey: <REDACTED>
  url: https://github.com/my-org
```

ApplicationSet example:
```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: applicationset-example
spec:
  generators:
    - git:
        repoURL: https://github.com/my-org/gitops-apps.git
        revision: develop
        directories:
        - path: applications/*/*
```

Running ApplicationSet v0.3.0

**Comment by bitgandtter** [comment]
Any insights? same issue here

**Comment by crenshaw-dev** [comment]
@agaudreault-jive this looks like a transient issue when the GitHub API fails to respond to a token refresh request within 15s. If it's super annoying, we could look at adding some retry logic.

**Comment by agaudreault-jive** [comment]
@crenshaw-dev if that is the case, a retry would fix the issue. I already had problems in another application using the same github app in concurrent process, so maybe that could be the source of error. 

> Make requests for a single user or client ID serially. Do not make requests for a single user or client ID concurrently.
https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-secondary-rate-limits.

I currently run argo with 5 repo-server pods.

**Comment by agaudreault-jive** [comment]
Looking deeper into this, we found an issue with our egress traffic configured for the nodes were some of the argoCD pods were running. The issue was causing request to take a lot longer than 15s. Since we moved the pods to nodes without the issue, I did not notice the problem anymore. Thanks @crenshaw-dev, your information helped me figure out the network problem.

I will close the issue, but if anyone has the same error message, it might be worth spending the time to make sure the network is behaving properly.
